azkaban-common/src/main/java/azkaban/executor/container/watch/AzPodStatusDriver.java

```diff
@@ -0,0 +1,143 @@
+    executor = Executors.newFixedThreadPool(THREAD_POOL_SIZE,
+        new ThreadFactoryBuilder().setNameFormat("azk-watch-pool-%d").build());
+
+    ImmutableMap.Builder listenerMapBuilder = ImmutableMap.builder()
+        .put(AzPodStatus.AZ_POD_REQUESTED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_SCHEDULED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_INIT_CONTAINERS_RUNNING, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_APP_CONTAINERS_STARTING, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_READY, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_COMPLETED, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_INIT_FAILURE, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_APP_FAILURE, new ArrayList<>())
+        .put(AzPodStatus.AZ_POD_UNEXPECTED, new ArrayList<>());
+    listenerMap = listenerMapBuilder.build();
+  }
+
+  /**
+   * Register a new {@link AzPodStatusListener}.
+   * Note that the callbacks are submitted to the {@link ExecutorService} in the order interfaces
+   * were registered, but that does not make any guarantees about the sequence in which all the
+   * callbacks for a given event are invoked.
+   *
+   * @param listener
+   */
+  public void registerAzPodStatusListener(AzPodStatusListener listener) {
+    requireNonNull(listener, "listener must not be null");
+    listenerMap.get(AzPodStatus.AZ_POD_REQUESTED).add(listener::onPodRequested);
+    listenerMap.get(AzPodStatus.AZ_POD_SCHEDULED).add(listener::onPodScheduled);
+    listenerMap.get(AzPodStatus.AZ_POD_INIT_CONTAINERS_RUNNING).add(listener::onPodInitContainersRunning);
+    listenerMap.get(AzPodStatus.AZ_POD_APP_CONTAINERS_STARTING).add(listener::onPodAppContainersStarting);
+    listenerMap.get(AzPodStatus.AZ_POD_READY).add(listener::onPodReady);
+    listenerMap.get(AzPodStatus.AZ_POD_COMPLETED).add(listener::onPodCompleted);
+    listenerMap.get(AzPodStatus.AZ_POD_INIT_FAILURE).add(listener::onPodInitFailure);
+    listenerMap.get(AzPodStatus.AZ_POD_APP_FAILURE).add(listener::onPodAppFailure);
+    listenerMap.get(AzPodStatus.AZ_POD_UNEXPECTED).add(listener::onPodUnexpected);
+
+  }
+
+  /**
+   * Shutdown the driver, including the {@link ExecutorService} with a timeout.
+   */
+  public void shutdown() {
+    executor.shutdown();
+    try {
+      executor.awaitTermination(SHUTDOW_TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+    } catch (InterruptedException e) {
+      logger.warn("Executor service shutdown was interrupted.", e);
+    }
+  }
+
+  /**
+   * Process all the callbacks for a given event. Lambdas corresponding to the callback method from
+   * every registered {@link AzPodStatusListener} are scheduled for execution on the {@code
+   * executor}
+   *
+   * @param podStatusMetadata
+   */
+  private void deliverCallbacksForEvent(AzPodStatusMetadata podStatusMetadata) {
+    listenerMap.get(podStatusMetadata.getAzPodStatus()).stream()
```

Comment:
The method `logPodWatchEvent` is already logging information for every incoming event, I am switching the log verbosity for it to `info`. 
In addition, as discussed, a more accurate metric for this will be the time spent by the callback tasks in the executor-service queue when a sufficient number of threads is not available. This will be added subsequently.